### PR TITLE
Exception creation, new testing, raising exception in retrieve, Fixes #3212

### LINKED ISSUE #3212: Background responses failures lack a stable code/name that maps to an exception class
### Confirm this is a feature request for the Python library and not the underlying OpenAI API.

- [x] This is a feature request for the Python library

### Describe the feature or improvement you're requesting

When the Responses API runs in background mode (`background=True`) and a run terminates in `status="failed"` (or `status="incomplete"`), the SDK does not surface a stable, machine-readable identifier that callers can map to an existing exception class.

The failure payload looks like:

```json
{
  "status": "failed",
  "error": { "code": "server_error", "message": "..." },
  "incomplete_details": null
}
```

The HTTP poll itself returns `200 OK`, so the SDK's status-code dispatcher is never invoked and **no typed exception is raised**. The only signal of *what kind of failure occurred* is the free-form `error.code` string, which:

- is **not documented as a stable enum**,
- does **not correspond 1:1 to any SDK exception class** (`server_error` ≠ `InternalServerError`, `rate_limit_exceeded` ≠ `RateLimitError`, etc.),
- is **not guaranteed to be present or non-null** for every failure mode (e.g. `incomplete` runs surface their reason in `incomplete_details.reason` instead, with yet another vocabulary),
- and may carry values not enumerated anywhere the SDK or callers can rely on.

### What exists today (synchronous mode)

For synchronous calls there is a clean, code-driven mapping in `openai/_client.py` keyed on HTTP status code:

```python
def _make_status_error(self, err_msg, *, body, response):
    data = body.get("error", body) if is_mapping(body) else body
    if response.status_code == 400: return BadRequestError(...)
    if response.status_code == 401: return AuthenticationError(...)
    if response.status_code == 403: return PermissionDeniedError(...)
    if response.status_code == 404: return NotFoundError(...)
    if response.status_code == 409: return ConflictError(...)
    if response.status_code == 422: return UnprocessableEntityError(...)
    if response.status_code == 429: return RateLimitError(...)
    if response.status_code >= 500: return InternalServerError(...)
    return APIStatusError(...)
```

So the synchronous path has a fully deterministic, documented contract: HTTP code → exception class. The background path has no equivalent — neither in the response body (no stable code) nor in the SDK (no dispatcher).

### Why this matters

In synchronous mode the SDK gives us a clean contract. In background mode the *same root cause* surfaces only as an opaque string with no contract. Any retry/backoff logic built around `isinstance(e, InternalServerError)` or `isinstance(e, RateLimitError)` simply cannot work for background runs, even though those are exactly the conditions where retry would be most useful.

### Suggested feature

Either:

1. **Add a stable, documented field** to background-run failure payloads (e.g. `error.type` or `error.class`) whose values directly correspond to existing SDK exception classes — e.g. `internal_server_error`, `rate_limit_error`, `bad_request_error`, `content_filter_error`, etc. The SDK can then mechanically map this field to the right exception class, mirroring what `_make_status_error` does today for HTTP status codes.

2. **Or: publish a documented, versioned mapping** from each possible `error.code` / `incomplete_details.reason` value to the SDK exception class (or "no class") and to a retryable flag. Treat it as part of the public API contract.

Either approach gives background runs the same deterministic contract synchronous calls already enjoy.

### Environment

- `openai` Python SDK version: `<fill in>`
- API: `POST /v1/responses` with `background=true`, polled via `GET /v1/responses/{id}`

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by alextemn**
## Changes being requested
-Adding response error to exception class mapping for async calls
-Creating exception from error response
-Raising an exception when status is "failed" or "incomplete"

## What to keep in mind
- Async jobs still return 200s, but we return appropriate exceptions
- If we want status codes to return the appropriate code, we'd need to make other, more upstream changes
- It also seems like synchronous jobs have the same issue of only returning 200s, since the server still gets back to us that there was an error, but I'm unsure if appropriate exceptions are raised, or if thats needed. I can open a new issue on this if needed

## Additional context & links
https://github.com/openai/openai-python/issues/3212
fixes #3212

### CODE REVIEW

File: src/openai/lib/__init__.py [hunk 1]
```diff
@@ -1,2 +1,7 @@
 from ._tools import pydantic_function_tool as pydantic_function_tool
 from ._parsing import ResponseFormatT as ResponseFormatT
+from ._responses import (
+    RESPONSE_ERROR_CODE_TO_EXCEPTION as RESPONSE_ERROR_CODE_TO_EXCEPTION,
+    INCOMPLETE_DETAILS_REASON_TO_EXCEPTION as INCOMPLETE_DETAILS_REASON_TO_EXCEPTION,
+    exception_for_background_failure as exception_for_background_failure,
```
[the remark this answers is not in the document]
Comment (alextemn) [reply src/openai/lib/__init__.py:6]: it is indeed a requirement to expose them to the SDK

**Review by jbeckwith-oai — CHANGES_REQUESTED** [review]
Thanks for putting this together. I appreciate the work here, especially the effort to cover both sync and async clients, think through raw/streaming wrappers, and add broad mapping tests. This is a tricky SDK boundary, and the underlying problem is real.

I reproduced the current behavior with a mocked `GET /responses/{id}` returning HTTP 200 and a body containing `status="failed"` plus `error.code="server_error"`. At the base commit, `responses.retrieve(...)` returns a normal `Response(status="failed")`; the SDK's HTTP 4xx/5xx exception dispatcher never runs. So there is a genuine usability gap here.

I'm requesting changes because I think we need to agree on the public SDK contract before this implementation can land. The issue is real; the open question is whether changing `retrieve()` into a throwing operation is the right solution.

## Code feedback

1. **The change makes `retrieve()` backward-incompatible.**

   `retrieve()` has historically meant “fetch and return the Response resource,” including failed, cancelled, and incomplete responses. This patch instead raises for selected terminal states, so existing polling code can no longer inspect `response.status`, `response.error`, partial output, or metadata. It also does not check `response.background`, so the new behavior applies to retrieved foreground responses too.

   The smallest/safest direction would be to keep default retrieval resource-returning and add an explicit classifier, `raise_for_status` helper, or opt-in throwing API. If automatic raising is desired, we should first explicitly agree that this compatibility change is acceptable and document the migration.

2. **The mapped exceptions have contradictory HTTP semantics.**

   `BadRequestError`, `RateLimitError`, and the other `APIStatusError` subclasses model HTTP 4xx/5xx responses. Here they are constructed from the successful polling response, so a `RateLimitError` can expose `status_code == 200` and `response.status_code == 200`.

   That is confusing for callers using status codes for retry logic, telemetry, or classification. A dedicated background-run failure/classification type would be clearer: it could preserve the full parsed `Response`, the original HTTP response/request ID, `error.code` or incomplete reason, and retryability without pretending the poll request returned 400/429/500.

3. **Unknown and missing failure data need a safer contract.**

   For `status="failed"` with no `error`, the helper returns `None`, so the failed response still looks like a normal return. That is one of the cases the issue calls out.

   Conversely, an unknown future `error.code` is guessed to be `InternalServerError(retryable=True)`. A future validation, policy, or quota code could be incorrectly treated as transient. I would use an explicit unknown background-failure classification with conservative/unknown retryability instead of guessing.

4. **The public behavior differs by wrapper, and the tests miss that boundary.**

   Plain `retrieve()` may raise, while `with_raw_response`, `with_streaming_response`, and `stream=True` bypass classification and return the failed resource. That may be a valid escape hatch, but it needs to be intentional and documented.

   The new tests are a good start, and the test file passes, but most tests patch private `_get()`. Please add sync and async public-client tests with a mocked HTTP 200 response, covering normal, raw, and streaming retrieval and asserting the public exception fields.

5. **Be careful about expanding the public API.**

   Re-exporting the mapping tables and helper from `openai.lib` makes them supported SDK surface area. The tables currently expose a private `_ErrorMapping` value type, and `.retryable` is an undeclared dynamic exception attribute. If the mapping is intended to be a stable public contract, it should be represented and documented as one; otherwise keep it internal.

## How to make the PR easier to review

The PR description currently lists implementation steps, but it does not make the problem and proposed contract clear enough for a reviewer who has not already read the issue.

A stronger description would include:

- a minimal reproduction;
- the exact before/after behavior;
- why raising from `retrieve()` is the right API instead of an explicit classifier/helper;
- the backward-compatibility impact;
- the source of truth for every error-code mapping;
- the intended meaning of `retryable`;
- explicit non-goals; and
- the tests you ran.

For example:

> Before: a background failure returns HTTP 200, and `responses.retrieve()` returns `Response(status="failed")`.
>
> After: [exact proposed behavior].
>
> This is safe because [compatibility explanation].
>
> Non-goals: [behavior this PR intentionally does not change].

The title could also describe the outcome rather than the implementation steps, for example: “Surface background Responses failures through a documented SDK error-classification API.”

Again, thank you for contributing. The problem is worth solving, and there is good work in the tests and edge-case thinking here. I think the next step is to nail down the public contract first, then adjust the implementation around that decision. Happy to follow up on the PR.

**Comment by alextemn** [comment]
@jbeckwith-oai Thanks for the detailed review — this all makes sense. I went through the options you laid out and agree the `raise_for_status` helper is the right one, so that's the direction I've taken.

The plan:

- **`retrieve()` goes back to how it was.** It returns the resource for every status, including `failed` and `incomplete`. That removes the backward-compatibility break, and with it the divergence between the plain, raw, and streaming wrappers.
- **A new `raise_for_status(response, *, raise_on_incomplete=False)` helper** in `openai.lib`. It raises on `status="failed"`, and on `status="incomplete"` only when the caller opts in. `cancelled` doesn't raise, since it's caller-initiated.
- **Dedicated exception types** instead of the `APIStatusError` subclasses, so nothing advertises a status code the poll didn't actually return. `ResponseFailedError` carries `.code`, `ResponseIncompleteError` carries `.reason`, and both sit on a shared `ResponseNotCompletedError` base holding the parsed `.response`, the `.request_id`, and a declared `.retryable`.

One thing I'd like your read on before I write the tests:

`raise_on_incomplete` defaults to `False`. A run that stops at `max_output_tokens` still leaves usable output on `response.output`, so raising on it by default felt surprising — happy to change that if you disagree.

**Comment by alextemn** [comment]
## Summary

A background Response reports its own outcome inside the body of the poll rather than in the HTTP status. When a
run fails, `GET /responses/{id}` still returns `200 OK`, so the SDK's 4xx/5xx exception dispatcher never runs and
`responses.retrieve()` hands back a `Response(status="failed")` that is easy to mistake for success.

This PR adds an opt-in classifier, `openai.lib.raise_for_status()`, plus three dedicated exception types. It does
not change the behavior of any existing method. Fixes #3212.

This replaces the earlier approach in this PR, which raised directly from `retrieve()`. See above for more detail

## Minimal reproduction

Mock `GET /responses/{id}` to return HTTP 200 with a terminal failure in the body — the same repro from the
review:

```python
import httpx2
from openai import OpenAI

body = {
    "id": "resp_123",
    "object": "response",
    "created_at": 1754925861,
    "model": "gpt-5",
    "output": [],
    "parallel_tool_calls": True,
    "tool_choice": "auto",
    "tools": [],
    "status": "failed",
    "background": True,
    "error": {"code": "server_error", "message": "boom"},
}

# the API answers the poll with 200 OK and reports the failure in the body
transport = httpx2.MockTransport(lambda request: httpx2.Response(200, json=body))
client = OpenAI(api_key="...", http_client=httpx2.Client(transport=transport))

response = client.responses.retrieve("resp_123")

print(response.status)  # "failed"
print(response.error.code)  # "server_error"
```

Nothing raises. A caller polling in a loop and checking only for exceptions treats this run as successful, and a
caller who does inspect `status` has no typed way to distinguish a transient infrastructure failure from a
permanent validation failure without string-matching `response.error.code` themselves.

## Before / After

**Before:** a background failure returns HTTP 200, and `responses.retrieve()` returns
`Response(status="failed")`. No exception is raised anywhere in the SDK. The `error` object is present on the
resource but unclassified, and `status="failed"` with a missing `error` object gives the caller nothing at all.

**After:** `responses.retrieve()` returns `Response(status="failed")`, exactly as before — unchanged. Callers who
want an exception opt in explicitly:

```python
from openai import OpenAI, ResponseFailedError
from openai.lib import raise_for_status

response = client.responses.retrieve("resp_123")

try:
    raise_for_status(response)
except ResponseFailedError as exc:
    print(exc.code)        # "server_error"
    print(exc.retryable)   # True
    print(exc.request_id)  # "req_123"
    print(exc.response)    # the parsed Response, still fully available
```

`raise_for_status(response, *, raise_on_incomplete: bool = False) -> None` behaves as follows:

| `response.status`       | Behavior                                                                              |
| ----------------------- | ------------------------------------------------------------------------------------- |
| `completed`             | returns                                                                               |
| `queued`, `in_progress` | returns — not a terminal state                                                        |
| `cancelled`             | returns — cancellation is caller-initiated, not a failure                             |
| `failed`                | raises `ResponseFailedError`                                                          |
| `incomplete`            | returns by default; raises `ResponseIncompleteError` when `raise_on_incomplete=True`  |

`incomplete` is gated behind an explicit opt-in because the common reason is `max_output_tokens`, which still
leaves usable content on `response.output`. Raising on that by default would turn a partial success into an
error for every caller who is happy to use the truncated output.

## Backward-compatibility impact

**None.** There is no diff in the actual `retrieve()` function:

```
 README.md                                     |  32 ++
 helpers.md                                    |  99 ++++
 src/openai/__init__.py                        |   6 +
 src/openai/_exceptions.py                     |  54 ++
 src/openai/lib/__init__.py                    |   1 +
 src/openai/lib/_responses.py                  |  46 ++
 tests/lib/test_responses_background_errors.py | 697 ++++++++++++++++++++++++++
```

Concretely:

- The new public surface is four names: `openai.lib.raise_for_status`, and `ResponseNotCompletedError`,
  `ResponseFailedError`, `ResponseIncompleteError` exported from the package root.
- The earlier revision re-exported the internal mapping tables from `openai.lib`; those are gone. `openai.lib`
  gains exactly one line.

## Source of truth for the error codes and statuses

There is no hand-maintained mapping table in this PR — that was a weakness of the first revision, which
duplicated all 20 error codes in a dict that would drift from the spec on every regeneration.

Classification now reads the response directly:

- `response.status` is matched against the two failure states. Every value comes from
  `openai.types.responses.response_status.ResponseStatus`.
- `ResponseFailedError.code` is `response.error.code`, passed through verbatim. It is never compared against a
  list of known codes, so an unrecognized code is preserved rather than coerced or dropped.
- `ResponseIncompleteError.reason` is `response.incomplete_details.reason`, likewise verbatim.

The only enumerated set in the implementation is the retryable-code set, which has three entries
(`_RETRYABLE_ERROR_CODES` in `src/openai/lib/_responses.py`).

The tests derive their parametrization from the generated models rather than hardcoding literals:

```python
ERROR_CODES      = get_args(ResponseError.model_fields["code"].annotation)          # all 20
INCOMPLETE_REASONS = get_args(...IncompleteDetails.model_fields["reason"]...)       # both
ALL_STATUSES     = get_args(ResponseStatus)                                         # all 6
```

So when the spec adds a code, status, or incomplete reason, coverage extends automatically instead of failing a
snapshot assertion. A separate test asserts the retryable set is a subset of the declared codes, which catches a
typo or a code removed upstream leaving a dead entry behind.

## The intended meaning of `retryable`

`retryable` answers one question: **is submitting a new `create()` call worth trying?**

It explicitly does **not** mean "call `retrieve()` again." A terminal response is immutable — re-polling the same
id returns the identical stored failure forever. This was raised on the first revision and the attribute is now
documented in those terms on the class itself, in `helpers.md`, and called out in a `> [!NOTE]` so it cannot be
read the other way.

`retryable is True` for exactly three transient codes: `server_error`, `rate_limit_exceeded`, and
`vector_store_timeout`.

Everything else is `False`, **including codes this version of the SDK has never seen.** The first revision
guessed unknown codes into `InternalServerError(retryable=True)`; a future validation, policy, or quota code
would then have been retried in a loop that could never succeed. Defaulting unknown codes to non-retryable fails
in the safe direction — a caller who misses a genuinely transient new code retries too little rather than
hammering the API.

`ResponseIncompleteError.retryable` is always `False`. Whether re-running a truncated generation is worthwhile
depends on the caller's `max_output_tokens` and prompt, which the SDK is not in a position to judge.

Two further contract details worth flagging for review:

- **`status="failed"` with no `error` object still raises**, with `code=None` and `retryable=False`. The first
  revision returned `None` here, which is one of the cases the original issue calls out.
- **The exceptions carry no HTTP status code.** They subclass `OpenAIError`, not `APIStatusError`. Constructing
  a `RateLimitError` from a successful poll would have produced an exception advertising `status_code == 200`
  while claiming to represent a 429. Callers keying retry, telemetry, or alerting off `status_code` would have
  been actively misled. `request_id` is preserved separately, from the poll's `x-request-id` header, so failed
  runs remain reportable.

## Non-goals

This PR intentionally does **not**:

- **Change `retrieve()`, or any other resource method.** Retrieval stays resource-returning for every status.
- **Make raising the default.** Classification is opt-in at the call site. Making it automatic is a separate
  decision that deserves its own discussion and a migration note.
- **Change HTTP status codes.** Background jobs still return 200 for in-band failures. Surfacing a real 4xx/5xx
  would require upstream API changes, not SDK changes.
- **Raise on `cancelled`.** Cancellation is caller-initiated and already observable; treating it as an error
  would be surprising.
- **Touch synchronous (foreground) response creation.** I noted earlier that `create()` may have a related
  gap where the server reports an error in a 200 body. It is out of scope here; happy to open a separate issue
  if it is worth tracking.
- **Add polling or `_and_poll` helpers for Responses.** Orthogonal, and better judged on its own merits.
- **Classify automatically inside `stream=True`.** The `response.failed` event carries the response; the caller
  passes it to the helper if they want an exception.

## Public API surface added

```python
from openai import (
    ResponseNotCompletedError,  # base: .response, .request_id, .retryable
    ResponseFailedError,        # + .code
    ResponseIncompleteError,    # + .reason
)
from openai.lib import raise_for_status
```

`ResponseNotCompletedError` exists so both cases can be caught in a single clause. Neither concrete type
subclasses the other.

## Tests

`tests/lib/test_responses_background_errors.py` — **116 cases across 39 test functions, all passing.**

No private attributes or methods are patched. The unit tests call the public helper directly; the end-to-end
tests drive the real `OpenAI` and `AsyncOpenAI` clients, with `respx2` mocking only the HTTP transport.

**Helper contract (74 cases)**

- Every one of the 20 declared error codes raises `ResponseFailedError` with `.code` and message preserved.
- Retryability asserted per code against the documented three-code set.
- Unknown future error code → raises, `.code` preserved verbatim, `retryable is False`.
- `failed` with `error=None` → raises, `code=None`, non-empty message.
- Unknown future incomplete reason and unknown future status handled conservatively.
- `incomplete` with no details, `{"reason": None}`, and `{}` all raise with `.reason is None` when opted in.
- `completed`, `cancelled`, `queued`, `in_progress`, and `status=None` never raise, even with
  `raise_on_incomplete=True`.
- A stale `error` object on a `completed` response does not trigger a raise — `status` is the authority.
- `background=True` and `background=False` classify identically.
- Exception hierarchy, absence of `status_code`, `APIStatusError` not in the MRO, `request_id` present and
  absent, and package-root exports.

**Public client, end to end (42 cases)**

Each parametrized over `normal` / `with_raw_response` / `with_streaming_response`, for **both** the sync and
async clients, against a mocked `200 OK` carrying the failure:

- Retrieval returns the resource through all three wrappers, for `failed` and `incomplete`.
- Full public exception surface asserted end to end: `str(exc)`, `.code`, `.retryable`, `.request_id`,
  `.response`, and the absence of `status_code`.
- Retryable and non-retryable codes both covered.
- `failed` with no `error` object.
- `incomplete` silent by default, raising on opt-in, on the same response object.
- `completed` and `cancelled` never raise over the wire.
- All three wrappers produce an identical exception, asserted side by side in one test.
- `stream=True` yields the `response.failed` event without raising; `event.response` fed to the helper raises.

Every test asserts the transport returned HTTP 200, so each one proves the exception arose *despite* a
successful poll rather than from HTTP error handling.

**Commands run**

```
pytest tests/lib/test_responses_background_errors.py   # 116 passed
pytest tests/lib/ tests/test_client.py                 # 1999 passed, 85 skipped
ruff check .                                           # clean
ruff format                                            # clean
pyright (strict)                                       # 0 errors, 0 warnings
mypy                                                   # 0 errors in changed files
```

I also mutation-tested the assertions that matter most, reverting each afterward, to confirm they are not
vacuous:

| Mutation                                          | Result      |
| ------------------------------------------------- | ----------- |
| Unknown codes → `retryable=True` (the old guess)  | 18 failed   |
| `failed` with `error=None` → silent return        | 1 failed    |
| `incomplete` raises by default                    | 5 failed    |
| `request_id` returns `None` unconditionally       | 26 failed   |

## Documentation

- `README.md` — a new subsection inside `## Handling errors`, placed right after the status-code table, so a
  reader who has just learned that 4xx/5xx raise `APIStatusError` immediately gets the caveat that this case
  does not.
- `helpers.md` — a new `# Response Status Helpers` section with the full status table, the exception attribute
  table, the `retryable` semantics including the re-`create()` vs re-`retrieve()` note, and async / raw /
  streaming / `stream=True` examples.

`api.md` is generated from the OpenAPI spec and `CHANGELOG.md` by release tooling, so neither is hand-edited
here.

@jbeckwith-oai
